### PR TITLE
feat(dot/sync): Implement warp sync strategy

### LINKED ISSUE #3183: Implement warp sync proof request
## Description

- Implement the warp sync proof request
- Implement the gap sync mechanism to download the chain state as we gather the warp sync fragments

### Additional informations

https://spec.polkadot.network/#sect-sync-warp

### PR DESCRIPTION
**Pull request by dimartiro**
## Changes

Implementation of the warp sync strategy.
This strategy is composed of 2 phases:
1. Request warp sync proofs until get one completed proof (last finalized block)
2. Retrieve that block to be used later for the state sync strategy

This goal of this strategy is to reach the latest finalized block of the canonical chain following authority set changes and validating justifications.
After this block is reached, we have to retrieve its state using the `state_sync`, that will allow us to build blocks on top of that one while we fill the gap between genesis and the latest finalized block retrieved by the warp sync.

### Summary of changes

- [x] Add new `warp_sync` strategy
   - [x] Add required warp sync p2p messages
   - [x] Add new reputation changes constants
- [x] Refactor peer set view to share the same reference between the strategies 
- [x] Add CLI config to use warp sync instead of full sync when node is started
- [x] Tests

## Tests

```sh
make test
```

## Issues

closes: #3183, #4277, #4268

### CODE REVIEW

File: dot/network/messages/warp_sync.go [hunk 1]
```diff
@@ -37,4 +46,37 @@ func (wpr *WarpProofRequest) String() string {
 	return fmt.Sprintf("WarpProofRequest begin=%v", wpr.Begin)
 }
 
+type WarpSyncFragment struct {
```
Comment (dimartiro) [inline dot/network/messages/warp_sync.go]: I had to recreate the same types we have in `lib/grandpa/warp_sync.go` to prevent a cyclic import.
Open to any suggestion to fix this 😄

File: dot/sync/warp_sync.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+const (
+	WarpProof = iota
+	TargetBlock
+	Completed
+)
+
+type WarpSyncStrategy struct {
+	// Strategy dependencies and config
+	peers            *peerViewSet
```
Comment (EclesioMeloJunior) [inline dot/sync/warp_sync.go:36]: the peer view set is important for `FullSync`, I am wondering if when WarpSync finishes the peerViewSet is lost or is it transferred to FullSync in some way?

Comment (dimartiro) [reply dot/sync/warp_sync.go:36]: I'm working on some changes in the `service.go` to handle the strategy switch and I'll consider this there! thks!

File: dot/sync/warp_sync.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+	logger.Infof("⛓️ synced %d warp sync fragments "+
+		"took: %.2f seconds, fps: %.2f fragments/second, target best block number #%d",
+		w.syncedFragments, totalSyncSeconds, fps, w.lastBlock.Number)
+}
+
+func (w *WarpSyncStrategy) IsSynced() bool {
+	return w.phase == Completed
+}
+
+func (w *WarpSyncStrategy) Result() any {
```
Comment (dimartiro) [inline dot/sync/warp_sync.go:289]: Will be used in the future as part of the `state_sync`

File: lib/common/hash.go [hunk 4]
```diff
@@ -62,7 +62,7 @@ func (h Hash) String() string { //skipcq: GO-W1029
 
 // Short returns the first 4 bytes and the last 4 bytes of the hex string for the hash
 func (h Hash) Short() string { //skipcq: GO-W1029
-	const nBytes = 4
+	const nBytes = 2
```
Comment (dimartiro) [inline lib/common/hash.go:65]: to show the same than substrate and easily compare hash in logs

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM! 🚀

File: dot/sync/service.go [hunk 5]
```diff
@@ -125,6 +127,13 @@ func NewSyncService(cfgs ...ServiceConfig) *SyncService {
 		cfg(svc)
 	}
 
+	// Set initial strategy
+	if svc.warpSyncStrategy != nil {
+		svc.currentStrategy = svc.warpSyncStrategy
+	} else {
+		svc.currentStrategy = svc.fullSyncStrategy
+	}
+
```
Comment (EclesioMeloJunior) [inline dot/sync/service.go:138]: Can we make it a `ServiceConfig`? `WithCurrentStrategy(s Strategy) { svc.currentStrategy = s } `

File: dot/sync/warp_sync.go (hunk 2 above)
Comment (EclesioMeloJunior) [reply dot/sync/warp_sync.go:36]: Have you found a way to pass those infos to the fullsync strategy?

Comment (dimartiro) [reply dot/sync/warp_sync.go:36]: Yes, if you check at the [service](https://github.com/ChainSafe/gossamer/pull/4275/files#diff-99bfbc3a2947ae495ea0c0fd19bcdcea2fa24a9d7fec49b4a9237361bb6709eaR522) we are sharing the same `peerViewSet` reference for all strategies

File: dot/sync/service.go (hunk 5 above)
Comment (dimartiro) [reply dot/sync/service.go:138]: I can add that option too, but I think this is something the service should decide not the caller

Comment (haikoschol) [reply dot/sync/service.go:138]: Doesn't the caller also decide this by calling `WithWarpSyncStrategy()`?

I don't see why it was necessary to move away from the `WithStrategies(current, default)` API. If the caller passes a `WarpSyncStrategy` instance as `current`, then that will be used initially, just as if they had used `WithWarpSyncStrategy()`.

Not that it matters much. It just seemed a bit more general without the service having member variables for specific types of strategies.

File: dot/sync/warp_sync.go [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+	WarpSyncProvider WarpSyncProofProvider
+	BlockState       BlockState
+	Peers            *peerViewSet
+}
+
+// NewWarpSyncStrategy returns a new warp sync strategy
+func NewWarpSyncStrategy(cfg *WarpSyncConfig) *WarpSyncStrategy {
+	authorities, err := cfg.WarpSyncProvider.CurrentAuthorities()
+	if err != nil {
+		panic("failed to get current authorities")
```
Comment (haikoschol) [inline dot/sync/warp_sync.go]: Maybe include the error in the panic message.

File: dot/sync/warp_sync.go [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+		badBlocks:        cfg.BadBlocks,
+		reqMaker:         cfg.RequestMaker,
+		peers:            cfg.Peers,
+		setId:            0,
+		authorities:      authorities,
+	}
+}
+
+// OnBlockAnnounce on every new block announce received
+// Synce it is a warp sync strategy, we are going to only update the peerset reputation
```
Comment (haikoschol) [inline dot/sync/warp_sync.go]:
```suggestion
// since(?) it is a warp sync strategy, we are going to only update the peerset reputation
```

File: dot/sync/warp_sync.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+		repChanges, bans, warpProofResult = w.validateWarpSyncResults(results)
+
+		if warpProofResult != nil {
+			if !warpProofResult.Completed {
+				logger.Debug("partial warp sync received")
+
+				// Partial warp proof
+				w.setId = warpProofResult.SetId
+				w.authorities = warpProofResult.AuthorityList
+				w.lastBlock = &warpProofResult.Header
```
Comment (haikoschol) [inline dot/sync/warp_sync.go]: This could be moved above the `if !warpProofResult.Completed` scope.

File: dot/sync/warp_sync.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+			}
+
+			// If invalid warp sync proof, then we should block the peer and update its reputation
+			encodedProof, err := response.Encode()
+			if err != nil {
+				// This should never happen since the proof is already decoded without issues
+				panic("fail to encode warp proof")
+			}
+
+			// Best proof will be the finished proof or the proof with more fragments
```
Comment (haikoschol) [inline dot/sync/warp_sync.go]: I guess this comment used to be above `if response.IsFinished || len(response.Proofs) > len(bestProof.Proofs)`.

IMHO it can be removed since that line very clearly expresses this condition.

File: lib/common/hash.go [hunk 10]
```diff
@@ -62,7 +62,7 @@ func (h Hash) String() string { //skipcq: GO-W1029
 
 // Short returns the first 4 bytes and the last 4 bytes of the hex string for the hash
```
Comment (haikoschol) [inline lib/common/hash.go]:
```suggestion
// Short returns the first 2 bytes and the last 2 bytes of the hex string for the hash
```

**Review by haikoschol — APPROVED** [review]
just a few nits/questions but LGTM!

File: dot/sync/service.go (hunk 5 above)
Comment (dimartiro) [reply dot/sync/service.go:138]: I'm thinking that could be better to get rid of the `WithWarpSyncStrategy` and `WithFullSyncStrategy` and let the sync service create this strategies by itself instead of relying on the caller.

File: dot/sync/warp_sync.go (hunk 7 above)
Comment (dimartiro) [reply dot/sync/warp_sync.go]: Yess thanks!

File: dot/sync/warp_sync.go (hunk 9 above)
Comment (dimartiro) [reply dot/sync/warp_sync.go]: Yess! thanks for the suggestion

File: dot/sync/service.go (hunk 5 above)
Comment (haikoschol) [reply dot/sync/service.go:138]: The downside of doing that is if some strategies have dependencies that the service doesn't need to know about, they could be passed directly to the strategy when instantiated, instead of passing them to the service. Not sure if that's currently the case or likely to be the case in the future though.

And we'd need another method of telling the service which strategy to use initially.

Comment (dimartiro) [reply dot/sync/service.go:138]: I think the strategy initialization and the strategy selection is part of the sync service internal implementation, the caller shouldn't know about it to prevent coupling our solution.

In terms of dependencies, the sync service should receive every dependency it (or the strategies) needs.

@timwu20 as your main architect, what are your thoughts about this?

File: dot/network/messages/warp_sync_test.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
...4c6a1faad6ef416aea872988d50227e35e2dfb5b73d4b3b897acaa4cf56f0ea674568468af9fef031f033166880eec9f5ddd05d797d5abf1a9b9a957c77820376d697522a9a42b9c6952e771fd74be9cb19fe6dd0d41c362f753ae2686547a2fdd6201a0e495db4830b2a4e80d045871b5f0f8541091fdb7707d5e6c8f085bc255f0a5069a0ae153d9a83f6fa3f208cd1d701fbd7e852449f026f8566c7d8826cf5b0aa99c0755eec29f2be753fb701762d1e7cc841323bac49576b9ea2e124c4b7b9f376d697522a9a42b9c6952e771fd74be9cb19fe6dd0d41c362f753ae2686547a2fdd62018c4f486b7b87c5735e42249c15f5ec6d7b2b50a16aa0ba3bfbe666f99537e81a6a121659a06d58eacb9594201b6644f18e54c1c6ca457a4fcfa9d15e3bfea201be0bbf54aad305682afc6dc7cd051b5dce154e7d0056e877562407872cdb1242376d697522a9a42b9c6952e771fd74be9cb19fe6dd0d41c362f753ae2686547a2fdd6201763fba0da4fef393f59f1fd9eb9a646a2a1106f404681840fa846ec56c1a3443223b8a147990c6fd3e2af11300531168be94f091931f6ac07724f3ce75a06600d3ccc285f3c648f34ca9ce476651495b9a6c4aa7607d83737bd871de56a50ed9376d697522a9a42b9c6952e771fd74be9cb19fe6dd0d41c362f753ae2686547a2fdd6201d9b7c0218e498e5930fc4ad9a8c139f774de3fda36ad408008c3957f1e822070ceda5ca810ef3fb7fda2c48576051107235edba92a3be39b38b9ac9aba8c2305f12e7ddae48873b7dcd614b8ccc8ea2e5ec000f3cc398f20801cf9047bc909b20001" //nolint:lll
```
Comment (timwu20) [inline dot/network/messages/warp_sync_test.go]: nit: move to a fixture?

File: dot/network/messages/warp_sync.go (hunk 1 above)
Comment (timwu20) [reply dot/network/messages/warp_sync.go]: Move `WarpSyncFragment` to another package and import in both packages?

File: dot/sync/service.go (hunk 5 above)
Comment (timwu20) [reply dot/sync/service.go:138]: I agree with Haiko, I don't really see a requirement to move away from the `WithStrategies(current, default)` API.  Seems to be quite strategy specific now by introducing `SyncService.fullSyncStrategy` and `SyncService.warpSyncStrategy` into the `SyncService` state.  Given that the initial `WithStrategies` call supported a current and default strategy, I would've assumed if we were to introduce a helper `WithWarpSync`method, it would just call `WithStrategies` with the actual strategies doing all the provider stuff within the method, instead of taking in the parameter of the strategy.

Comment (dimartiro) [reply dot/sync/service.go:138]: I strongly believe that keeping the strategy selection and initialization within the sync service’s internal implementation is the better approach, and here’s why:

- **Better encapsulation**: by having the sync service handle its own strategy setup and selection, we avoid coupling the selection logic to external callers. This allows the service to evolve internally without requiring changes from every caller whenever we add or adjust a strategy. While this might mean we need to provide additional dependencies when creating the sync service, that’s due to the lack of a more comprehensive dependency injection mechanism.

- **Avoiding exposure of internal implementations**: strategy selection should be an implementation detail managed by the sync service, not something callers need to be aware of or control. Callers should only need to interact with the service without knowing the internal details of how it chooses or initializes the strategy—or even that we’re using "strategies" at all.

- **Open to change**: If we decide to include more than two strategies (like the `stateSyncStrategy` I'm currently working on), or if we restructure how strategies are used, having the sync service manage its own strategy logic will make these changes significantly easier. It allows us to avoid adjustments in each external caller and its tests, keeping the interface simpler and more resilient to internal changes.

In short, managing the strategies internally not only keeps the design cleaner but also future-proofs it.

Apologies if I sound overly firm, but I genuinely believe this is the right path forward. That said, if after considering these points you still feel this approach is flawed, I'm happy to go along with the current implementation 😄

File: dot/sync/service.go [hunk 12]
```diff
@@ -92,6 +92,7 @@ type Strategy interface {
 	Process(results []*SyncTaskResult) (done bool, repChanges []Change, blocks []peer.ID, err error)
 	ShowMetrics()
 	IsSynced() bool
+	Result() any
```
Comment (timwu20) [inline dot/sync/service.go:95]: why do we need this? I don't see it called anywhere.

File: dot/sync/service.go (hunk 5 above)
Comment (timwu20) [reply dot/sync/service.go:138]: I guess I don't really see the reason to change much of this existing code.  I attempted to try and see how it would have looked without these changes (see 5b0207b190d6b165fd6675c03bda93b537e5ce5e), and it seems inline with the approach that @EclesioMeloJunior authored. 

I would give @EclesioMeloJunior final authority here.  He already approved this PR, but I would like to hear what he has to say on this thread.

File: dot/sync/service.go (hunk 12 above)
Comment (dimartiro) [reply dot/sync/service.go:95]: This will be used by the state sync in the future. I can remove it from this PR and include it in a new one if you want

File: dot/sync/service.go (hunk 5 above)
Comment (EclesioMeloJunior) [reply dot/sync/service.go:138]: > I think the strategy initialization and the strategy selection is part of the sync service internal implementation, the caller shouldn't know about it to prevent coupling our solution.

That's not true for polkadot-sdk, basically the caller can decide what strategy the sync should use, that decision is also made via cli `--warp-sync`. 

So, I would say that we should stick with `currentStrategy` and `defaultStrategy` and enable the definitions of both to the caller.

With regards the points @dimartiro put:
- Better encapsulation: I disagree on that since we should enable the caller/user choose the desired sync strategy, that is easy due to the usage of interfaces which enables a less encapsulation (we don't depend on the concrete impl, since we can just provide whatever satisfies the `Strategy` interface) 
- Avoiding exposure of internal implementations: Here is the point, we should not expose the implementation to the part that is using it, which means that the sync engine should have exposure just for the interface methods. Each strategy should be self contained, the caller is the one who provides us one valid implementation, you can imagine sync as a modular system where you can plug and un-plug strategies at runtime.
- Open to change: If we hardcode that warp sync is the default strategy then we will might face problems to sync if we start from scratch or if we start a single node, what I am suggesting is make it as less coupled as we can where we leave the choice to the caller/user, so depending on its needs.

Comment (haikoschol) [reply dot/sync/service.go:138]: @dimartiro makes a good point regarding the state sync strategy. I know we discussed this before, but I don't remember how we decided to handle this. IIRC, state sync needs to run concurrently to full sync after warp sync has completed. In our previous discussion I think we mentioned two options:

1. Drive state sync from `FullSyncStrategy`

For this `FullSyncStrategy` needs to be informed that it also needs to run state sync, in the case where warp sync ran to completion before `FullSyncStrategy` is started. I assume @dimartiro intends to use the [added `Result()` method](https://github.com/ChainSafe/gossamer/pull/4275#discussion_r1852981665) in the `Strategy` interface for this somehow.

2. Drive state sync from `SyncService`

For this, the service needs to figure out when it needs to run state sync strategy in addition to `FullSyncStrategy`. That would require keeping more than two strategies in the service and changing the API discussed in this thread as well.

As I mentioned before, I don't think it matters that much since I don't see us adding many new sync strategies in the future.

Comment (dimartiro) [reply dot/sync/service.go:138]: @EclesioMeloJunior Regarding this:

> That's not true for polkadot-sdk, basically the caller can decide what strategy the sync should use, that decision is also made via CLI --warp-sync.

We need to differentiate between two aspects here. On one hand, we have the `sync mode`, which could be either `warp` or `full` sync. On the other hand, we have the implementation, i.e., the strategies.  
In my opinion, the caller shouldn't need to know about the strategies but only about the `sync mode`. The sync mode is a parameter required by the sync service, but the internal implementation, including decisions about where and when to configure, instantiate, and use strategies, should be the responsibility of the sync service, not the caller.  

For instance, suppose we decide to change the approach and stop using strategies in favor of another solution. Even if we currently have an interface in place, we would still need to modify the caller to adapt to the new approach. My main point is to avoid this by encapsulating all the sync logic within the sync service.

Regarding what Substrate does: the sync service is responsible for creating the strategies, not the caller. You can take a look [here](https://github.com/paritytech/polkadot-sdk/blob/6d5f8141ad42d7a528a09715c492974550709b92/substrate/client/network/sync/src/strategy/polkadot.rs#L355-L391).

As @haikoschol pointed out, we are going to need to use "state sync" in the future. Are we expecting the caller to configure, instantiate, and send the 3 strategies to the sync service instead of letting the sync service decide which strategy to instantiate and use?  

In my opinion, that's not the right architectural decision for our system. However, I’m open to accepting your approach and leaving it as is if you prefer. I just wanted to raise these concerns beforehand.

Comment (EclesioMeloJunior) [reply dot/sync/service.go:138]: @haikoschol `state sync` should not run in parallel with `full sync` after `warp sync finishes`, what happens is that `warp sync` finishes and then starts `state sync` that should be completed before starting `full sync`. What should happens "together" with full sync is the `gap sync` which is this case will be part of full sync and not a strategy under its own.

@dimartiro yeah, I saw the changes and maybe then we can do is:

```go
// Set initial strategy
func WithWarpSyncStrategy(warpSyncStrategy Strategy) ServiceConfig {
	return func(svc *SyncService) {
		svc.warpSyncStrategy = warpSyncStrategy
                 svc.currentStrategy = warpSyncStrategy
	}
}

func WithFullSyncStrategy(fullSyncStrategy Strategy) ServiceConfig {
	return func(svc *SyncService) {
		svc.fullSyncStrategy = fullSyncStrategy
                svc.currentStrategy = fullSyncStrategy
	}
}
```

the you don't need the if/else, since when the strategy is defined by the config you already set it as current.

Comment (haikoschol) [reply dot/sync/service.go:138]: @EclesioMeloJunior thanks for clarifying that!

IMHO your proposed changes make the behavior very confusing. Looking at the two function signatures, I would never expect that the order in which I call them matters or that they overwrite `svc.currentStrategy`.

Comment (dimartiro) [reply dot/sync/service.go:138]: Let's proceed as follows: since this discussion is taking too many iterations and blocking this PR, I'd prefer to use the simpler version with `current` and `default` for now. However, I believe we'll need to revisit this discussion in the next PR when I introduce the state sync.

File: dot/network/messages/warp_sync_test.go (hunk 11 above)
Comment (dimartiro) [reply dot/network/messages/warp_sync_test.go]: Are we using fixtures in any other place? 🤔

Comment (haikoschol) [reply dot/network/messages/warp_sync_test.go]: Looks like we do in [`dot/sync/testdata/westend_blocks.yaml`](https://github.com/ChainSafe/gossamer/blob/a46ad52411cb95dad9e706c8c725f00ff940f36d/dot/sync/testdata/westend_blocks.yaml).

Comment (dimartiro) [reply dot/network/messages/warp_sync_test.go]: Thank you @haikoschol ! I addressed this [here](https://github.com/ChainSafe/gossamer/pull/4275/commits/474fb6c74c1f10ab4c52e58aadfb19eececf9aaa)

File: dot/network/messages/warp_sync.go (hunk 1 above)
Comment (dimartiro) [reply dot/network/messages/warp_sync.go]: Done!
